dir.go

```diff
@@ -403,9 +403,15 @@ func (d *Directory) sync() error {
 		if err != nil {
 			return err
 		}
+
+		// Although cache is cleared at end of function, delete each entry
+		// after update to prevent unnecessary future update in case of failure
+		// before end of loop
+		delete(d.entriesCache, name)
```

Comment:
This is unsafe, IIRC, if we have open files. That's why we have a separate "this is totally unsafe only call this if you know what you're doing" `Uncache` function.

TL;DR: the cache isn't a cache, it's a record of open files.

Reply:
_Please provide more context and expand on your argument._

We are already clearing the cache after the loop. Is _this_ particularity line 'unsafe', or any cache modification inside this function?

> the cache isn't a cache, it's a record of open files.

What does this mean? This structure functions like a cache. The `Uncache` function is just an ugly optimization to prevent cache bloating: https://github.com/ipfs/go-ipfs/pull/3027#issuecomment-448678709. `Uncache` deletes a cache entry, same as `Unlink` (we won't flush an entry we are removing) and `sync` here.

Reply:
> We are already clearing the cache after the loop. Is this particularity line 'unsafe', or any cache modification inside this function?

Where? As far as I can tell, this PR _adds_ the cache clearing after the loop (which has the same problem).

> What does this mean? This structure functions like a cache.

https://github.com/ipfs/go-mfs/issues/54?

Reply:
Yes, we've been abusing this. I still think removing the entry here is no more evil than what is happening elsewhere, but if this is just a performance optimization we can skip it.

Reply:
It is an attempt to reduce memory use, not so much for performance.  It may actually decrease performance by having to retrieve a node from the DAG, instead of getting it from the cache.

After the above discussion, I have stopped clearing `entriesCache` in `sync()`.

Reply:
> Yes, we've been abusing this. I still think removing the entry here is no more evil than what is happening elsewhere, but if this is just a performance optimization we can skip it.

It's bad, but there we _know_ it's safe because we own the MFS object. Here, it's definitely unsafe (concurrent `ipfs files *` operations will race and conflict).